tests/test_nodes.py

```diff
@@ -560,3 +568,87 @@ def test_allow_objects() -> None:
     iv = IllegalType()
     c.foo = iv
     assert c.foo == iv
+
+
+@pytest.mark.parametrize(  # type: ignore
+    "node, value, expected",
+    [
+        (UnionNode(ref_type=Union[int, str], element_types=[int, str], value=1), 1, 1),
+        (UnionNode(element_types=[int, str], value="str"), "str", "str"),
+        (UnionNode(element_types=[int, str]), "str", "str"),
+        (UnionNode(element_types=[int, str, type(None)]), None, None),
+        (
+            UnionNode(element_types=[DictConfig, str]),
+            DictConfig({"a": 1}),
+            DictConfig({"a": 1}),
+        ),
+        (
+            UnionNode(element_types=[ListConfig, int]),
+            ListConfig([1, 2]),
+            ListConfig([1, 2]),
+        ),
+        (UnionNode(element_types=[List[int], int]), [1, 2], [1, 2]),
+        (UnionNode(element_types=[List[bool], bool]), [True, False], [True, False]),
+        (UnionNode(element_types=[List[float], float]), [3.14, 9.67], [3.14, 9.67]),
+        (UnionNode(element_types=[List[str], str]), ["foo", "foo2"], ["foo", "foo2"]),
+        (
+            UnionNode(element_types=[List[User], str]),
+            [User(name="user1", age=21), User(name="user2", age=45)],
+            [User(name="user1", age=21), User(name="user2", age=45)],
+        ),
+        (
+            UnionNode(element_types=[List[int], int]),
+            1,
+            1,
+        ),
+        (UnionNode(element_types=[Dict[str, int], int]), {"foo": 1}, {"foo": 1}),
+        (
+            UnionNode(element_types=[Dict[str, str], str]),
+            {"foo": "var"},
+            {"foo": "var"},
+        ),
+        (
+            UnionNode(element_types=[Dict[str, bool], bool]),
+            {"foo": True},
+            {"foo": True},
+        ),
+        (
+            UnionNode(element_types=[Dict[str, float], float]),
+            {"foo": 3.14},
+            {"foo": 3.14},
+        ),
+        (
+            UnionNode(element_types=[bool, float], is_optional=True),
```

Comment:
I'll look into it

Reply:
Done, ref_type stays as Union[NoneType, X], elements_types is without Nonetype 'Union[x]' and is_optional is True if NoneType is inside ref_type.

Reply:
So actually, I just suggestd to forbid `Union[NoneType, X]` to keep things simple.
@omry do you see a good reason to support it?

Reply:
Union[X] is exactly the same as Union[NoneType, X]:
```
In [1]: from typing import Union, Optional

In [2]: class X: ...

In [3]: Optional[X]
Out[3]: typing.Union[__main__.X, NoneType]

In [4]: Union[X, type(None)] == Optional[X]
Out[4]: True

In [5]: Union[type(None), X] == Optional[X]
Out[5]: True
```

optional fields has special handling in OmegaConf (because their support predated Union support) but in theory it could be implemented in terms of a Union instead.

Reply:
> Optional[X] is exactly the same as Union[NoneType, X]:

Yes, that's why I suggested to support only `Optional[X]` and not `Union[NoneType, X]`: there is no need to support both.

Although we could probably support both, it makes things a bit more complex because now we have a `Union` that shouldn't translated into a `UnionNode` (it should instead be an `XNode` with `optional=True`).

Keeping this conversation resolved but feel free to reopen if you disagree.

Reply:
Actually, unresolving this conversation because I just realized that in Python, `Optional[X]` is actually translated into `Union[NoneType, X]`. Thus I assume that OmegaConf is already catching this kind of Union somewhere to turn it into an `XNode` with `optional=True`. As a result, the error triggered at https://github.com/pereman2/omegaconf/blob/7d6e39ebed85279ef9cae6cd91bcb77748b2df22/omegaconf/_utils.py#L457 should never happen => it probably should be an assert (I don't think there was a test for it, actually).